### PR TITLE
feat: allow non-tedge users to own and run tedge

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Allow non-tedge users to own and run tedge.

* allow users to control the user/group via the system.toml
* support running everything as the "current user" (when setting empty user/group values in the system.toml)
* all commands should respect these settings, e.g. `tedge init`, `c8y-remote-access-plugin --init`, `tedge cert` etc.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/3528

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s. You can activate automatic signing by running `just prepare-dev` once)
- [x] I ran `just format` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `just check` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: configuration/package_scripts/tedge/preinst [hunk 1]
```diff
@@ -29,31 +29,51 @@ user_exists() {
...
+        fi
     fi
 fi
 
-### Create file in /etc/sudoers.d directory. With this configuration, the tedge user have the right to call the tedge command with sudo rights, which is required for system-wide configuration in "/etc/tedge"
-if [ -d /etc/sudoers.d ]; then
+### Create file in /etc/sudoers.d directory. With this configuration,
+###the configured user has the right to call the tedge command with sudo rights,
+### which is required for system-wide configuration in "/etc/tedge"
+if [ "$TEDGE_USER" = "tedge" ] && [ -d /etc/sudoers.d ]; then
```
Comment (reubenmiller) [inline configuration/package_scripts/tedge/preinst:92]: I'm wondering if we really need the "tedge" specific check here? Couldn't we just reference the `$TEDGE_USER` variable within the sudoers.d file creation instead of using the hard coded "tedge" value?

File: configuration/package_scripts/tedge/preinst [hunk 2]
```diff
@@ -29,31 +29,51 @@ user_exists() {
...
 
+# Read user/group from /etc/tedge/system.toml if it exists
+TEDGE_USER="tedge"
+TEDGE_GROUP="tedge"
+SYSTEM_TOML="/etc/tedge/system.toml"
+if [ -f "$SYSTEM_TOML" ]; then
+    _val=$(grep -E '^[[:space:]]*user[[:space:]]*=' "$SYSTEM_TOML" | head -1 | sed 's/.*=[[:space:]]*"\(.*\)".*/\1/')
+    [ -n "$_val" ] && TEDGE_USER="$_val"
+    _val=$(grep -E '^[[:space:]]*group[[:space:]]*=' "$SYSTEM_TOML" | head -1 | sed 's/.*=[[:space:]]*"\(.*\)".*/\1/')
+    [ -n "$_val" ] && TEDGE_GROUP="$_val"
```
Comment (reubenmiller) [inline configuration/package_scripts/tedge/preinst]: Generally we use `head -n1` instead of `head -1` as the former is more compatible.

The grep and sed are pretty fancy, but I guess we can live without given that we don't have any good toml parsing tools available.

File: configuration/package_scripts/tedge/preinst (hunk 1 above)
Comment (albinsuresh) [reply configuration/package_scripts/tedge/preinst:92]: Yeah, I was also not fully convinced what the "secure practice" here is. Initially I had it as you're suggesting, esp for convenience (less things for him to do). But then I was wondering if granting sudo permissions to a user that we didn't create is okay or not, from a security point of view. Ideally, the user should be managing his own permissions, right? Or is it okay for a tool that he installs to make those changes on his behalf? 

But may be, we can argue that we're only giving the user the necessary sudo rights to run `tedge` tools only and nothing more.

File: tests/RobotFramework/tests/installation/install_tedge.robot [hunk 3]
```diff
@@ -49,10 +55,57 @@ Install then uninstall latest tedge via script (from main branch)
...
+    Transfer To Device    ${CURDIR}/system.toml    /etc/tedge/system.toml
+
+    # Install tedge
+    Install tedge
+
+    Execute Command    cmd=sed -i s/User=tedge/User=tedge-custom/ /lib/systemd/system/tedge-agent.service
+    Execute Command    cmd=sed -i s/User=tedge/User=tedge-custom/ /lib/systemd/system/tedge-mapper-c8y.service
+    Execute Command    cmd=systemctl daemon-reload
+
+    Connect device    ${DEVICE_SN}
```
Comment (albinsuresh) [inline tests/RobotFramework/tests/installation/install_tedge.robot]: Something weird is happening here. Before `tedge connect` is executed here, that starts the `tedge-agent`, the service definition has `User=tedge-custom` as I have set in the previous steps. But after `tedge connect` that is getting switched back to `User=tedge`. That's weird and I wonder why. Investigating.

Comment (albinsuresh) [reply tests/RobotFramework/tests/installation/install_tedge.robot]: Damn, that was a silly dev error. The `Connect device` call, instead of just connecting the device, was re-executing the boostrap script as well, reinstalling `tedge` again, wiping that systemd definition changes made in the previous step.

Comment (albinsuresh) [reply tests/RobotFramework/tests/installation/install_tedge.robot]: Resolved by [c738cbf](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c738cbf0840d9d820eb584e64a95d9256e51f6b7)

File: configuration/package_scripts/tedge/preinst (hunk 1 above)
Comment (reubenmiller) [reply configuration/package_scripts/tedge/preinst:92]: Though these permissions can be critical to some of the tedge functionality, e.g. if we can't use the tedge-write command, then the config plugin will not work in a lot of cases. If the system.toml is set to root permissions, then a non-root user can't change these values anyway.

Adding a check for the "tedge" user specifically feels a bit arbitrary, as the user would also get created automatically (which is normally an action reserved for root users)

Comment (reubenmiller) [reply configuration/package_scripts/tedge/preinst:92]: ah sorry I misread the part in the postinst script regarding the user creation! I thought it was creating the users automatically, and that's why I thought if were were doing that, then why not set the sudoers rules!

So ignore my comment. Sorry for the confusion

File: configuration/package_scripts/tedge/preinst (hunk 2 above)
Comment (reubenmiller) [reply configuration/package_scripts/tedge/preinst]: We'll also need to improve the snippet regarding parsing the system.toml as it assumes a well-formed file (e.g. if i forget to have closing quotes, `user = "foo`, then the awk still matches, and results in an unexpected value.

And `[ -n "$_val" ] && TEDGE_GROUP="$_val"` will also stop the postinst execution if there is no value matched due to the usage of `set -e` at the top of the script. We can just use a plain if statement rather than using `<test> && true || false` logic.

Though I can refine this logic, and check all the corner case (if you want).

File: crates/core/tedge/src/cli/init.rs [hunk 4]
```diff
@@ -76,12 +92,10 @@ impl TEdgeInitCmd {
...
-        let permissions = {
-            PermissionEntry::new(
-                Some(self.user.clone()),
-                Some(self.group.clone()),
-                Some(0o775),
-            )
+        let permissions = if user.is_empty() && group.is_empty() {
+            PermissionEntry::new(None, None, Some(0o775))
+        } else {
+            PermissionEntry::new(Some(user.clone()), Some(group.clone()), Some(0o775))
```
Comment (rina23q) [inline crates/core/tedge/src/cli/init.rs]: I think the logic is incomplete. It's missing to address one of `user` or `group` is empty.

```suggestion
        let permissions = PermissionEntry::new(
            (!user.is_empty()).then(|| user.clone()),
            (!group.is_empty()).then(|| group.clone()),
            Some(0o775),
        );
```

File: crates/core/tedge/src/cli/init.rs [hunk 5]
```diff
@@ -103,14 +117,46 @@ impl TEdgeInitCmd {
         create_directory_and_update_ownership(&config.logs.path, &permissions).await?;
         create_directory_and_update_ownership(&config.data.path, &permissions).await?;
 
-        let entity_store_file = config_dir.join(".agent").join("entity_store.jsonl");
+        let system_toml = config_dir.join("system.toml");
+        if system_toml.exists() && !user.is_empty() && !group.is_empty() {
```
Comment (rina23q) [inline crates/core/tedge/src/cli/init.rs]: Why does it limit to the case **both** `user` and `group` exist?

File: crates/core/tedge_mapper/src/c8y/mapper.rs [hunk 6]
```diff
@@ -405,8 +406,11 @@ pub async fn bridge_rules(
 ) -> anyhow::Result<BridgeConfig> {
     let mapper_config_dir =
         tedge_config.mapper_config_dir::<C8yMapperSpecificConfig>(cloud_profile);
+    let system_config = SystemConfig::try_new(tedge_config.root_dir()).unwrap_or_default();
+    let (tedge_user, tedge_group) = (system_config.user.as_str(), system_config.group.as_str());
     if let Err(err) =
-        create_directory_with_user_group(mapper_config_dir.clone(), "tedge", "tedge", 0o755).await
+        create_directory_with_user_group(mapper_config_dir.clone(), tedge_user, tedge_group, 0o755)
```
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y/mapper.rs]: And here there is no guard to check if `tedge_user.is_empty()` or `tedge_group.is_empty()`. So, the directory creation will fail in that case.

**Review by rina23q — COMMENTED** [review]
I have minor comments regarding how to address a corner case where toml file has `user` or `group` key but empty value.

File: crates/core/tedge/src/cli/connect/command.rs [hunk 7]
```diff
@@ -951,21 +951,28 @@ impl ConnectCommand {
...
     if bridge_config.auth_type == AuthType::Basic {
         return;
     }
 
     let (user, group) = match bridge_config.bridge_location {
-        BridgeLocation::BuiltIn => ("tedge", "tedge"),
-        BridgeLocation::Mosquitto => (crate::BROKER_USER, crate::BROKER_GROUP),
+        BridgeLocation::BuiltIn => {
+            let system_config =
+                tedge_config::SystemConfig::try_new(tedge_config.root_dir()).unwrap_or_default();
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/connect/command.rs]: This deserves a `TedgeConfig` method

```suggestion
               tedge_config.read_system_config();
```

File: crates/core/tedge/src/cli/mod.rs [hunk 8]
```diff
@@ -97,13 +97,17 @@ pub enum Component {
 pub enum TEdgeOpt {
     /// Initialize Thin Edge
     Init {
-        /// The user who will own the directories created
-        #[clap(long, default_value = "tedge")]
-        user: String,
+        /// The user who will own the directories created.
+        /// Overrides the value from system.toml [user].
+        /// Defaults to "tedge".
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/mod.rs]: This is not accurate. `tedge` is the default only if no user is provided by `system.toml`

File: crates/core/tedge/src/cli/mod.rs [hunk 9]
```diff
@@ -97,13 +97,17 @@ pub enum Component {
...
+        #[clap(long)]
+        user: Option<String>,
 
-        /// The group who will own the directories created
-        #[clap(long, default_value = "tedge")]
-        group: String,
+        /// The group who will own the directories created.
+        /// Overrides the value from system.toml [group].
+        /// Defaults to "tedge".
```
Comment (didier-wenzek) [inline crates/core/tedge/src/cli/mod.rs]: As for `--user`, the default value is not accurate.

File: crates/core/tedge_mapper/src/c8y/mapper.rs [hunk 10]
```diff
@@ -405,8 +406,11 @@ pub async fn bridge_rules(
 ) -> anyhow::Result<BridgeConfig> {
     let mapper_config_dir =
         tedge_config.mapper_config_dir::<C8yMapperSpecificConfig>(cloud_profile);
+    let system_config = SystemConfig::try_new(tedge_config.root_dir()).unwrap_or_default();
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/mapper.rs]:
```suggestion
    let system_config = tedge_config.read_system_config();
```

Comment (albinsuresh) [reply crates/core/tedge_mapper/src/c8y/mapper.rs]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: crates/core/tedge/src/cli/mod.rs (hunk 9 above)
Comment (albinsuresh) [reply crates/core/tedge/src/cli/mod.rs]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: crates/core/tedge/src/cli/mod.rs (hunk 8 above)
Comment (albinsuresh) [reply crates/core/tedge/src/cli/mod.rs]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: crates/core/tedge/src/cli/connect/command.rs (hunk 7 above)
Comment (albinsuresh) [reply crates/core/tedge/src/cli/connect/command.rs]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: crates/core/tedge_mapper/src/c8y/mapper.rs (hunk 6 above)
Comment (albinsuresh) [reply crates/core/tedge_mapper/src/c8y/mapper.rs]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: crates/core/tedge/src/cli/init.rs (hunk 4 above)
Comment (albinsuresh) [reply crates/core/tedge/src/cli/init.rs]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: configuration/package_scripts/tedge/preinst (hunk 2 above)
Comment (albinsuresh) [reply configuration/package_scripts/tedge/preinst]: Resolved by [c22b8f8](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/c22b8f878b574e75af8a156f40e310b0986f5559)

File: crates/core/tedge/src/cli/init.rs (hunk 5 above)
Comment (albinsuresh) [reply crates/core/tedge/src/cli/init.rs]: Resolved by [0c56812](https://github.com/thin-edge/thin-edge.io/pull/4094/commits/0c56812a46185c3fc1ccb7dd578915147be7147a)

File: configuration/package_scripts/tedge/preinst [hunk 11]
```diff
@@ -29,31 +29,66 @@ user_exists() {
...
+
+# Read user/group from /etc/tedge/system.toml if it exists
+TEDGE_USER="tedge"
+TEDGE_GROUP="tedge"
+SYSTEM_TOML="/etc/tedge/system.toml"
+if [ -f "$SYSTEM_TOML" ]; then
+    _val=$(parse_toml_string_value "user" "$SYSTEM_TOML")
+    TEDGE_USER="${_val:-$TEDGE_USER}"
+    _val=$(parse_toml_string_value "group" "$SYSTEM_TOML")
+    TEDGE_GROUP="${_val:-$TEDGE_GROUP}"
```
Comment (reubenmiller) [inline configuration/package_scripts/tedge/preinst]: There is a slight logic error here when the system.toml has empty values, e.g.

```sh
user = ""
group = ""
```

But I can apply the fix for this.

File: crates/common/tedge_config/src/tedge_toml/tedge_config_location.rs [hunk 12]
```diff
@@ -470,7 +471,10 @@ impl TEdgeConfigLocation {
 
         atomically_write_file_async(toml_path, toml.as_bytes()).await?;
 
-        if let Err(err) = change_user_and_group(toml_path, "tedge", "tedge").await {
+        let system_config = SystemConfig::try_new(&self.tedge_config_root_path).unwrap_or_default();
+        if let Err(err) =
+            change_user_and_group(toml_path, &system_config.user, &system_config.group).await
```
Comment (reubenmiller) [inline crates/common/tedge_config/src/tedge_toml/tedge_config_location.rs]: I suspect we could move the guard around the empty user and/or group condition to inside the `change_user_and_group` function (technically inside the private `change_user_and_group_sync` function).

As doing so would ensure that any parts of the code that use that function directly (which there are a few places), would get the same handling.

Examples of direct calls to the `change_user_and_group`:
* https://github.com/albinsuresh/thin-edge.io/blob/9c464e6444c1a017eae04428ad0c86833688bb9a/crates/core/tedge/src/cli/connect/command.rs#L974
* https://github.com/albinsuresh/thin-edge.io/blob/9c464e6444c1a017eae04428ad0c86833688bb9a/plugins/c8y_remote_access_plugin/src/lib.rs#L107

File: configuration/package_scripts/tedge/preinst [hunk 13]
```diff
@@ -29,31 +29,67 @@ user_exists() {
...
+    if [ -n "$MATCHES" ]; then
+        # property is defined, so respect all values, even empty ones
+        echo "$MATCHES" | sed 's/.*=[[:space:]]*"\([^"]*\)".*/\1/'
     else
-        echo "WARNING: Could not create group: tedge" >&2
+        # not-defined so the given default value
+        echo "Using default as no valid value for '$property' was found in '$file_path'" >&2
+        echo "$default_value"
+    fi
+}
```
Comment (Bravo555) [inline configuration/package_scripts/tedge/preinst:48]: note: this could fail if in the future we were to add `user` or `group` key under any other table (or for any other key if there's more than one)

File: crates/core/tedge/src/cli/init.rs [hunk 14]
```diff
@@ -103,14 +102,52 @@ impl TEdgeInitCmd {
...
+        } else {
+            let agent_state_dir = config_dir.join(".agent");
+            create_directory_and_update_ownership(&agent_state_dir, &permissions).await?;
+            agent_state_dir
+        };
 
-        if entity_store_file.exists() {
-            change_user_and_group(entity_store_file, &self.user, &self.group).await?;
+        let entity_store_file = agent_state_dir.join("entity_store.jsonl");
+        if entity_store_file.exists() && !user.is_empty() && !group.is_empty() {
```
Comment (Bravo555) [inline crates/core/tedge/src/cli/init.rs]: suggestion: don't have to check if `user`/`group` are empty because it's already done by `PermissionEntry::apply`.

```suggestion
        if entity_store_file.exists() {
```

File: crates/core/tedge/src/cli/certificate/cli.rs [hunk 15]
```diff
@@ -229,9 +230,10 @@ pub enum CA {
 impl BuildCommand for TEdgeCertCli {
     async fn build_command(self, config: &TEdgeConfig) -> Result<Box<dyn Command>, ConfigError> {
         let (user, group) = if config.mqtt.bridge.built_in {
-            ("tedge", "tedge")
+            let system_config = SystemConfig::try_new(config.root_dir()).unwrap_or_default();
```
Comment (rina23q) [inline crates/core/tedge/src/cli/certificate/cli.rs]: It looks forgotten to change:

```suggestion
            let system_config = config.read_system_config();
```

File: crates/common/tedge_utils/src/file.rs [hunk 16]
```diff
@@ -399,6 +399,18 @@ impl PermissionEntry {
     }
 }
 
+/// Creates a `PermissionEntry` from plain `user` and `group` strings.
+///
+/// Empty strings are treated as unset and mapped to `None`, so ownership
+/// is left unchanged for whichever field is empty.
+pub fn permissions(user: &str, group: &str, mode: u32) -> PermissionEntry {
```
Comment (rina23q) [inline crates/common/tedge_utils/src/file.rs:406]: (trivial) I think this function deserves staying under `PermissionEntry` struct, however, it's fine.

File: tests/RobotFramework/tests/installation/install_tedge.robot [hunk 17]
```diff
@@ -49,10 +55,41 @@ Install then uninstall latest tedge via script (from main branch)
...
+    ...    before installation forces tedge to use that user for all of its file ownership.
+
+    Create Custom User And Group
+    Install Tedge With System Config    empty_user_system.toml
+
+    # Set the desired tedge user manually, instead of specifying it in system.toml
+    Execute Command    tedge init --user ${CUSTOM_USER} --group ${CUSTOM_GROUP}
+
+    Setup Systemd Overrides For Custom User
+    Verify Tedge Running With Custom User
```
Comment (Bravo555) [inline tests/RobotFramework/tests/installation/install_tedge.robot]: question: Not sure exactly what this is testing.

I assume the description is out of date and what it _should_ say instead is that we check that when user/group is empty then installation shouldn't change current ownership. But to test that we'd have to have some previous files and after install verify owner didn't change. But here we're doing `tedge init` which explicitly overwrites the owner.

File: tests/RobotFramework/tests/customizing/tedge_init.robot [hunk 18]
```diff
@@ -67,10 +68,113 @@ Tedge init and check if default values are restored
...
+    Execute Command    sudo c8y-remote-access-plugin --init
+    All Tedge Directories Should Be Owned By    tedge:tedge
+    [Teardown]    Restore Initial State
+
+Tedge init leaves ownership unchanged when user and group are empty in system.toml
+    Transfer To Device    ${CURDIR}/resources/empty_user_empty_group_system.toml    /etc/tedge/system.toml
+    Execute Command    sudo tedge init
+    Execute Command    sudo c8y-remote-access-plugin --init
+    All Tedge Directories Should Be Owned By    root:root
+    [Teardown]    Restore Initial State
```
Comment (Bravo555) [inline tests/RobotFramework/tests/customizing/tedge_init.robot:90]: question: By "unchanged" we mean that `init` commands recreate the files so their new owner becomes whatever user runs the `init` command, and **not** that existing files do not have their owner changed?

Comment (reubenmiller) [reply tests/RobotFramework/tests/customizing/tedge_init.robot:90]: In this case, generally an explicit "empty" user/group means that the process can't change ownership of files/folders anyway, so attempting to would be anyway pointless.

So yes, new files should then have the owner set to the current processes owner...but I think this happens  without explicitly doing anything.

File: tests/RobotFramework/tests/installation/install_tedge.robot (hunk 17 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/installation/install_tedge.robot]: It looks to be testing two things, the pre-seeding (e.g. making sure the preinst part of the package doesn't try to apply the "tedge" user), ~and the second is to check that the `--user` and `--group` flags override what ever is set in the system.toml~

update:

Looking at it again, it is testing a corner case where the user wants to handle the entire initialization themselves whilst still using the debian package to install thin-edge.io. The empty values in the system.toml should control the preinst and postinst logic...though tbh this test case doesn't provide much value, as this is not the recommended sequence. If you want to use a custom user, then just set explicit values in the system.toml...if you want to run as the current user, then you should not be using the debian package.

File: configuration/package_scripts/tedge/preinst (hunk 13 above)
Comment (reubenmiller) [reply configuration/package_scripts/tedge/preinst:48]: This can't happen as we're piping to ` | head -n1` (after the grep) ;)

Comment (reubenmiller) [reply configuration/package_scripts/tedge/preinst:48]: But yes, the parsing is not ideal, but we can live with it since we don't generally add new properties to the system.toml, and until we have a more generic toml cli commands, then we can't do anything else (the downside of the toml format tbh).

File: crates/common/tedge_utils/src/file.rs (hunk 16 above)
Comment (reubenmiller) [reply crates/common/tedge_utils/src/file.rs:406]: I can't judge tbh, so if this is to be done, then someone else will need to push the fix.

File: crates/core/tedge/src/cli/init.rs (hunk 14 above)
Comment (reubenmiller) [reply crates/core/tedge/src/cli/init.rs]: Resolved by 62b36c459b3258b3cb3debb29bae3e9ee11b2b17

File: crates/core/tedge/src/cli/certificate/cli.rs (hunk 15 above)
Comment (reubenmiller) [reply crates/core/tedge/src/cli/certificate/cli.rs]: Resolved by 714ba74268011d69f990fbf77172372031a2f23c

File: crates/common/tedge_utils/src/file.rs (hunk 16 above)
Comment (rina23q) [reply crates/common/tedge_utils/src/file.rs:406]: I would say keep it as it is for now. I would rather like to suggest a refactoring of this file after a release.

File: tests/RobotFramework/tests/customizing/tedge_init.robot [hunk 19]
```diff
@@ -67,10 +68,113 @@ Tedge init and check if default values are restored
...
+
+Tedge init defaults both user and group to tedge when system.toml has an invalid user value
```
Comment (rina23q) [inline tests/RobotFramework/tests/customizing/tedge_init.robot:92]: I am a bit surprised to see that we force to change `tedge:tedge` when `system.toml` contains invalid user or group value.

```toml
user = 42
group = "tedge"
```

I think it would be better if it doesn't change anything and throws an error, so that users can aware of their mistake and modify the wrong `system.toml`.

File: tests/RobotFramework/tests/installation/install_tedge.robot (hunk 17 above)
Comment (rina23q) [reply tests/RobotFramework/tests/installation/install_tedge.robot]: I don't think this test is useful either, since the behaviour when defining empty values for `user`/`group` are well covered in `customizing/tedge_init.robot`.

**Review by rina23q — APPROVED** [review]
I think the code and test coverage are good. I left a comment, but it doesn't intend to block the PR. I just wanted to express my opinion.

File: tests/RobotFramework/tests/customizing/tedge_init.robot (hunk 19 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/customizing/tedge_init.robot:92]: Your suggestion might be more correct, however given the lack of a formal toml parser on the command line, we won't be able to do this easily, and generally the `tedge init` will fail when using tedge:tedge as the user won't exist (hence why the user was editing the user/group in the system.toml in the first place).

Comment (reubenmiller) [reply tests/RobotFramework/tests/customizing/tedge_init.robot:92]: so I'd just suggest leaving it as is for now

File: tests/RobotFramework/tests/installation/install_tedge.robot (hunk 17 above)
Comment (reubenmiller) [reply tests/RobotFramework/tests/installation/install_tedge.robot]: Removed in 9f82209e72c075c37dbc1acf8831f99fe2258f50
